File: srcjs/modules/draw.js

const drawHLine = (opts) => {
  let yintercept = opts.yintercept || 0;
  let color = opts.color || "#000";
  let width = opts.width || 1;
  let dash = opts.dash || [];
  const func = (u) => {
    let y = u.valToPos(yintercept, "y", true);
    let xl = u.valToPos(u.scales.x.min, "x", true);
    let xr = u.valToPos(u.scales.x.max, "x", true);
    u.ctx.save();
    u.ctx.beginPath();
    u.ctx.moveTo(xl, y);
    u.ctx.lineTo(xr, y);
    u.ctx.strokeStyle = color;
    u.ctx.lineWidth = width;
    u.ctx.setLineDash(dash);
    u.ctx.stroke();
    u.ctx.restore();
  };
  return func;
};


const drawVLine = (opts) => {
  let xintercept = opts.xintercept || 0;
  let color = opts.color || "#000";
  let width = opts.width || 1;
  let dash = opts.dash || [];
  const func = (u) => {
    let x = u.valToPos(xintercept, "x", true);
    let yb = u.valToPos(u.scales.y.min, "y", true);
    let yt = u.valToPos(u.scales.y.max, "y", true);
    u.ctx.save();
    u.ctx.beginPath();
    u.ctx.moveTo(x, yt);
    u.ctx.lineTo(x, yb);
    u.ctx.strokeStyle = color;
    u.ctx.lineWidth = width;
    u.ctx.setLineDash(dash);
    u.ctx.stroke();
    u.ctx.restore();
  };
  return func;
};


const drawVRect = (opts) => {
  let xmin = opts.xmin || 0;
  let xmax = opts.xmax || 0;
  let fill = opts.fill || "#848484";
  let alpha = opts.alpha || 0.1;
  const func = (u) => {
    let yt = u.valToPos(u.scales.y.max, "y", true);
    let yb = u.valToPos(u.scales.y.min, "y", true);
    let xl = u.valToPos(xmin, "x", true);
    let xr = u.valToPos(xmax, "x", true);
    u.ctx.save();
    u.ctx.beginPath();
    u.ctx.fillStyle = fill;
    u.ctx.globalAlpha = alpha;
    u.ctx.fillRect(xl, yt, xr - xl, yb - yt);
    u.ctx.restore();
  };
  return func;
};


const drawHRect = (opts) => {
  let ymin = opts.ymin || 0;
  let ymax = opts.ymax || 0;
  let fill = opts.fill || "#848484";
  let alpha = opts.alpha || 0.1;
  const func = (u) => {
    let yb = u.valToPos(ymin, "y", true);
    let yt = u.valToPos(ymax, "y", true);
    let xl = u.valToPos(u.scales.x.min, "x", true);
    let xr = u.valToPos(u.scales.x.max, "x", true);
    u.ctx.save();
    u.ctx.beginPath();
    u.ctx.fillStyle = fill;
    u.ctx.globalAlpha = alpha;
    u.ctx.fillRect(xl, yt, xr - xl, yb - yt);
    u.ctx.restore();
  };
  return func;
};


export { drawHLine, drawVLine, drawVRect, drawHRect };

File: R/draw.R


#' Hooks for drawing lines and rectangles
#'
#' @param uplot Chart created with [uPlot()].
#' @param xintercept,yintercept,xmin,xmax,ymin,ymax Coordinates for lines and rectangles.
#' @param color,fill Color.
#' @param width Width for lines.
#' @param dash Line dash pattern.
#' @param alpha Transparency parameter.
#'
#' @return An `htmlwidget` object of class `"uPlot"`.
#' @export
#'
#' @name draw-lines-rectangles
#'
#' @example examples/ex-draw.R
uHookVLine <- function(uplot, xintercept, color = NULL, width = NULL, dash = NULL) {
  xintercept <- if (inherits(xintercept, "Date")) {
    as.numeric(xintercept) * 86400
  } else {
    as.numeric(xintercept)
  }
  check_uplot(uplot)
  params <- drop_nulls(list(
    xintercept = xintercept,
    color = color,
    width = width,
    dash = dash
  ))
  uplot$x$config$options$hooks$drawAxes <- c(
    uplot$x$config$options$hooks$drawAxes,
    list(
      JS(sprintf("drawVLine(%s)", jsonlite::toJSON(params)))
    )
  )
  return(uplot)
}

#' @export
#'
#' @rdname draw-lines-rectangles
uHookHLine <- function(uplot, yintercept, color = NULL, width = NULL, dash = NULL) {
  yintercept <- if (inherits(yintercept, "Date")) {
    as.numeric(yintercept) * 86400
  } else {
    as.numeric(yintercept)
  }
  check_uplot(uplot)
  params <- drop_nulls(list(
    yintercept = yintercept,
    color = color,
    width = width,
    dash = dash
  ))
  uplot$x$config$options$hooks$drawAxes <- c(
    uplot$x$config$options$hooks$drawAxes,
    list(
      JS(sprintf("drawHLine(%s)", jsonlite::toJSON(params)))
    )
  )
  return(uplot)
}

#' @export
#'
#' @rdname draw-lines-rectangles
uHookVRect <- function(uplot, xmin, xmax, fill = NULL, alpha = NULL) {
  xmin <- if (inherits(xmin, "Date")) {
    as.numeric(xmin) * 86400
  } else {
    as.numeric(xmin)
  }
  xmax <- if (inherits(xmax, "Date")) {
    as.numeric(xmax) * 86400
  } else {
    as.numeric(xmax)
  }
  check_uplot(uplot)
  params <- drop_nulls(list(
    xmin = xmin,
    xmax = xmax,
    fill = fill,
    alpha = alpha
  ))
  uplot$x$config$options$hooks$drawAxes <- c(
    uplot$x$config$options$hooks$drawAxes,
    list(
      JS(sprintf("drawVRect(%s)", jsonlite::toJSON(params)))
    )
  )
  return(uplot)
}


#' @export
#'
#' @rdname draw-lines-rectangles
uHookHRect <- function(uplot, ymin, ymax, fill = NULL, alpha = NULL) {
  ymin <- if (inherits(ymin, "Date")) {
    as.numeric(ymin) * 86400
  } else {
    as.numeric(ymin)
  }
  ymax <- if (inherits(ymax, "Date")) {
    as.numeric(ymax) * 86400
  } else {
    as.numeric(ymax)
  }
  check_uplot(uplot)
  params <- drop_nulls(list(
    ymin = ymin,
    ymax = ymax,
    fill = fill,
    alpha = alpha
  ))
  uplot$x$config$options$hooks$drawAxes <- c(
    uplot$x$config$options$hooks$drawAxes,
    list(
      JS(sprintf("drawHRect(%s)", jsonlite::toJSON(params)))
    )
  )
  return(uplot)
}

